### ISSUE TITLE
ERROR: Could not build wheels for sparse-dot-topn which use PEP 517 and cannot be installed directly

### ISSUE THREAD

**Issue by dt1086** [ing-bank/sparse_dot_topn#60]
Hi - I'm encountering issues trying to install sprase_dot_topn.

My version of Python is 3.8.5. Below is the error message any ideas?

```
ERROR: Command errored out with exit status 1:
   command: 'C:\Users\17178\anaconda3\envs\learn-env\python.exe' 'C:\Users\17178\anaconda3\envs\learn-env\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py' build_wheel 'C:\Users\17178\AppData\Local\Temp\tmpy9eo9hr0'
       cwd: C:\Users\17178\AppData\Local\Temp\pip-install-uyb94kqo\sparse-dot-topn_2defdd3130d34138b0d3e3deb199880e
  Complete output (42 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.8
  creating build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\awesome_cossim_topn.py -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\__init__.py -> build\lib.win-amd64-3.8\sparse_dot_topn
  running egg_info
  writing sparse_dot_topn.egg-info\PKG-INFO
  writing dependency_links to sparse_dot_topn.egg-info\dependency_links.txt
  writing requirements to sparse_dot_topn.egg-info\requires.txt
  writing top-level names to sparse_dot_topn.egg-info\top_level.txt
  reading manifest file 'sparse_dot_topn.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  writing manifest file 'sparse_dot_topn.egg-info\SOURCES.txt'
  copying sparse_dot_topn\array_wrappers.pxd -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\array_wrappers.pyx -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn.pyx -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn_parallel.cpp -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn_parallel.h -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn_source.cpp -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn_source.h -> build\lib.win-amd64-3.8\sparse_dot_topn
  copying sparse_dot_topn\sparse_dot_topn_threaded.pyx -> build\lib.win-amd64-3.8\sparse_dot_topn
  running build_ext
  cythoning ./sparse_dot_topn/array_wrappers.pyx to ./sparse_dot_topn\array_wrappers.cpp
  cythoning ./sparse_dot_topn/sparse_dot_topn.pyx to ./sparse_dot_topn\sparse_dot_topn.cpp
  cythoning ./sparse_dot_topn/sparse_dot_topn_threaded.pyx to ./sparse_dot_topn\sparse_dot_topn_threaded.cpp
  building 'sparse_dot_topn.array_wrappers' extension
  creating build\temp.win-amd64-3.8
  creating build\temp.win-amd64-3.8\Release
  creating build\temp.win-amd64-3.8\Release\sparse_dot_topn
  C:\Users\17178\anaconda3\envs\learn-env\Library\mingw-w64\bin\gcc.exe -mdll -O -Wall -DMS_WIN64 -IC:\Users\17178\anaconda3\envs\learn-env\include -IC:\Users\17178\anaconda3\envs\learn-env\include -IC:\Users\17178\AppData\Local\Temp\pip-build-env-mqor7oz3\normal\Lib\site-packages\numpy\core\include -c ./sparse_dot_topn\array_wrappers.cpp -o build\temp.win-amd64-3.8\Release\.\sparse_dot_topn\array_wrappers.o -Ox
  cc1plus.exe: error: argument to '-O' should be a non-negative integer, 'g', 's' or 'fast'
  C:\Users\17178\AppData\Local\Temp\pip-build-env-mqor7oz3\normal\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\17178\AppData\Local\Temp\pip-install-uyb94kqo\sparse-dot-topn_2defdd3130d34138b0d3e3deb199880e\sparse_dot_topn\array_wrappers.pxd
    tree = Parsing.p_module(s, pxd, full_module_name)
  C:\Users\17178\AppData\Local\Temp\pip-build-env-mqor7oz3\normal\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\17178\AppData\Local\Temp\pip-install-uyb94kqo\sparse-dot-topn_2defdd3130d34138b0d3e3deb199880e\sparse_dot_topn\sparse_dot_topn.pyx
    tree = Parsing.p_module(s, pxd, full_module_name)
  C:\Users\17178\AppData\Local\Temp\pip-build-env-mqor7oz3\normal\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\17178\AppData\Local\Temp\pip-install-uyb94kqo\sparse-dot-topn_2defdd3130d34138b0d3e3deb199880e\sparse_dot_topn\sparse_dot_topn_threaded.pyx
    tree = Parsing.p_module(s, pxd, full_module_name)
  error: command 'C:\\Users\\17178\\anaconda3\\envs\\learn-env\\Library\\mingw-w64\\bin\\gcc.exe' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for sparse-dot-topn
ERROR: Could not build wheels for sparse-dot-topn which use PEP 517 and cannot be installed directly
```

**Comment by junotown** [comment]
Hi, I was just wondering if you fixed this error? I am getting the same error

**Comment by dt1086** [comment]
Nope, issue still persists.

**Comment by mmlynarik** [comment]
I've had the same problem when running python from docker image python:3.8.5-slim. When I changed it into python:3.8.5, it started to work.

**Comment by Darenar** [comment]
On my Ubuntu machine with python 3.8.5, the issue was resolved by:
`apt-get update`
`apt-get install -y pkg-config`
`pip install sparse_dot_topn`

If it does not work - try updating pip and setup tools also
`pip install --upgrade pip setuptools wheel`

**Comment by sophie-DataEngineer** [comment]
@Darenar  Could you please share your Dockerfile, thank you so much!

**Comment by vdtdg** [comment]
Hi everyone, I have the same problem when installing the library from an Airflow 2.2.0 container.

Here is my dockerfile:
```
FROM apache/airflow:2.2.0

USER root
RUN apt-get update
RUN apt-get install -y pkg-config

RUN pip install --upgrade pip
RUN pip install wheel
RUN pip install setuptools>=42
RUN pip install cython>=0.29.15
RUN pip install scipy>=1.2.3
RUN pip install numpy>=1.16.6

RUN pip install sparse_dot_topn
...
```

Here is the Airflow container which I derive from : [https://github.com/apache/airflow/blob/main/Dockerfile](https://github.com/apache/airflow/blob/main/Dockerfile)

Airflow container is based on the Python base docker images. As @mmlynarik, maybe it comes from them building it from a slim image ? But then, what could we run in the Dockerfile to fill what might be missing ?

To reproduce, simple try to build the dockerfile and see the same error mentionned by @dt1086.

**Comment by gtechzilla** [comment]
What worked for me was to ensure the numpy  version is 1.19.5 and string-grouper version is 0.6.1

**Comment by ymwdalex** [comment]
@vdtdg @sophie-DataEngineer @Darenar @mmlynarik @dt1086 @junotown 

Thanks for involving the discussion of this thread! We just released [v0.3.3](https://pypi.org/project/sparse-dot-topn/#files). We employed Github actions and CiBuildWheel, and make the wheels available for Py3.6-3.10 on multiple architectures of Linux, MacOS and Windows.

If you still have installation problems, please try the latest version.

If you find any problem with v0.3.3, please let us know. Thanks!

**Comment by ymwdalex** [comment]
[v0.3.3](https://pypi.org/project/sparse-dot-topn/#files) should fix this issue
